### PR TITLE
Implement Threadmark support

### PR DESCRIPTION
**Pull request by jrahmy**
Major things left:

- [x] On display rounded word count totals, not the exact.
- [X] Display wordcount on threadmark preview (hover over next/previous links)
- [x] Display total wordcount in non-overlay mode for the threadmark index.
- [x] Pick a consistent method for determining if threadmark functionality is installed/active that respects if the threadmarks add-on is disabled.
- [x] Thread search results displaying total word-count for threads with threadmarks

### CODE REVIEW

**Comment by Xon** [comment]
So major things left:

- On display rounded word count totals, not the exact.
- Pick a consistent method for determining if threadmark functionality is installed/active that respects if the threadmarks add-on is disabled.

**Comment by Xon** [comment]
Pushed the things left into the PR description, and pushed 2 commits.

File: upload/library/SV/WordCountSearch/XenForo/Model/Thread.php [hunk 1]
```diff
@@ -49,15 +19,17 @@ public function getThreadmarkWordCountByThread($threadId, $cache = true)
...
-            );
-        }
+        return $wordCount;
+    }
+
+    public function rebuildThreadWordCount($threadId)
+    {
+        $wordCount = $this->countThreadmarkWordsInThread($threadId);
+
+        $dw = XenForo_DataWriter::create('XenForo_DataWriter_Discussion_Thread');
```
Comment (jrahmy) [inline upload/library/SV/WordCountSearch/XenForo/Model/Thread.php]: Too much overhead? Should it be a simple query instead?

**Comment by jrahmy** [comment]
I believe I've taken care of the outstanding issues.

There is something strange happening though. If I rebuild the search index, _all_ post word counts are being indexed, not just those above the threshold. Looking at the current master (https://github.com/Xon/XenForo-WordCountSearch/blob/master/upload/library/SV/WordCountSearch/XenForo/Search/DataHandler/Post.php#L7-L18), it seems to me that the threshold is only checked for inserting the row into the database, but it's still pushed into the metadata array for indexing regardless? Is this the desired behavior? I just want to make sure I'm not introducing a bug here.

File: upload/library/SV/WordCountSearch/Installer.php [hunk 2]
```diff
@@ -47,4 +69,16 @@ public static function uninstall()
             DROP TABLE IF EXISTS `xf_post_words`
         ");
     }
-}
\ No newline at end of file
+
+    public static function addOnIsActive($addOnId)
```
Comment (jrahmy) [inline upload/library/SV/WordCountSearch/Installer.php]: Maybe a good candidate for https://github.com/Xon/XenForo-Utils

File: upload/library/SV/WordCountSearch/XenForo/Model/Thread.php (hunk 1 above)
Comment (Xon) [inline upload/library/SV/WordCountSearch/XenForo/Model/Thread.php]: Should be fine, This is a fairly rare operation and how it resorts a threads a ludicrously large overhead anyway.

Comment (Xon) [inline upload/library/SV/WordCountSearch/XenForo/Model/Thread.php]: This should be fine. Rebuilding thread positional data is such a ludicrously large overhead this isn't even noticeable.

File: addon-SV_WordCountSearch.xml [hunk 3]
```diff
@@ -79,8 +80,8 @@ $0]]></replace>
     </modification>
     <modification template="search_result_thread" modification_key="SV_WordCountSearch_search_result_thread_1" description="Inject Word count" execution_order="10" enabled="1" action="preg_replace">
       <find><![CDATA[#\w*{xen:phrase thread_by}#si]]></find>
-      <replace><![CDATA[<xen:if is="{$post.WordCount}">
-{xen:phrase word_count}: <a class="wordcount">{$post.WordCount}</a>,
+      <replace><![CDATA[<xen:if is="{$thread.word_count}">
+{xen:phrase word_count}: <a class="wordcount">{$thread.word_count}</a>,
```
Comment (Xon) [inline addon-SV_WordCountSearch.xml]: This should be a rounded threadcount like posts are.

**Comment by Xon** [comment]
@jrahmy it is intended behaviour that the word count is always pushed into the search index (so you can search it), but not stored in the database unless it is above the threshold.

This makes ~500mb of difference for the database on SV, and nearly gb or so on SB. I'm actually considering pushing the rounded wordcount into the search index to better match Elastic's Search  inverted index behaviour.

**Comment by jrahmy** [comment]
>@jrahmy it is intended behaviour that the word count is always pushed into the search index (so you can search it), but not stored in the database unless it is above the threshold.
Oh okay, it all makes much more sense to me now (go figure). For whatever reason, I thought posts weren't meant to be searchable by word count unless above the threshold.

**Pull request merged**